File: src/rust/flows_development/src/handoff.rs
//! Automatic handoff: git commit of the target directory (excluding `.harness/`) + a line
//! in `progress.txt`, without spending a model turn. A failure at any step falls back to
//! the legacy manual-repair prompt (`prompts::handoff_prompt`).

use std::path::{Path, PathBuf};

use harness_engine::{feature_store, git_command, run_config_store, state_store};

use crate::prompts;
use crate::tasks::{CURRENT_FEATURE_ID_KEY, CURRENT_FEATURE_SUMMARY_KEY, CURRENT_FEATURE_TITLE_KEY};

fn state(key: &str) -> String {
    state_store::get(key).unwrap_or_default()
}

pub fn complete_verified_feature(verify_result: &str) -> String {
    match try_automated_handoff(verify_result) {
        Ok(confirmation) => {
            eprintln!("[dev] automatic handoff completed: {confirmation}");
            if let Ok(id) = state(CURRENT_FEATURE_ID_KEY).parse::<i32>() {
                feature_store::mark_passed(id);
            }
            if feature_store::all_passing() {
                crate::tasks::done()
            } else {
                prompts::bearings_prompt()
            }
        }
        Err(failure) => {
            eprintln!("[dev] automatic handoff failed: {failure}");
            prompts::handoff_prompt(Some(&failure))
        }
    }
}

fn try_automated_handoff(verify_result: &str) -> Result<String, String> {
    let feature_id: i32 = state(CURRENT_FEATURE_ID_KEY)
        .parse()
        .map_err(|_| "current feature missing from state.json".to_string())?;

    let feature = feature_store::load()
        .into_iter()
        .find(|f| f.id == feature_id);
    let mut title = feature
        .map(|f| f.title)
        .unwrap_or_else(|| state(CURRENT_FEATURE_TITLE_KEY));
    if title.trim().is_empty() {
        title = format!("feature #{feature_id}");
    }

    let config = run_config_store::load();
    let target_dir = resolve_target_dir(&config.target_dir)?;

    std::fs::create_dir_all(&target_dir)
        .map_err(|e| format!("failed to update progress.txt: {e}"))?;
    append_progress(
        &target_dir,
        feature_id,
        &title,
        &config.verify_cmd,
        verify_result,
    )
    .map_err(|e| format!("failed to update progress.txt: {e}"))?;

    let rev_parse = git_command::run(&target_dir, &["rev-parse", "--show-toplevel"]);
    if rev_parse.exit_code != 0 {
        return Ok(format!(
            "NO_GIT: {}",
            one_line(
                &rev_parse.error,
                "target directory is outside a Git repository"
            )
        ));
    }

    let add = git_command::run(&target_dir, &["add", "-A", "--", ".", ":(exclude).harness"]);
    if add.exit_code != 0 {
        return Err(format!(
            "git add failed: {}",
            one_line(&add.error, &add.output)
        ));
    }

    let diff = git_command::run(
        &target_dir,
        &[
            "diff",
            "--cached",
            "--quiet",
            "--",
            ".",
            ":(exclude).harness",
        ],
    );
    if diff.exit_code == 0 {
        let head = git_command::run(&target_dir, &["rev-parse", "--short", "HEAD"]);
        return Ok(if head.exit_code == 0 {
            one_line(&head.output, "NO_CHANGES")
        } else {
            "NO_CHANGES".to_string()
        });
    }
    if diff.exit_code > 1 {
        return Err(format!(
            "git diff --cached failed: {}",
            one_line(&diff.error, &diff.output)
        ));
    }

    let commit = git_command::run(
        &target_dir,
        &[
            "commit",
            "-m",
            &commit_message(feature_id, &title),
            "--",
            ".",
            ":(exclude).harness",
        ],
    );
    if commit.exit_code != 0 {
        return Err(format!(
            "git commit failed: {}",
            one_line(&commit.error, &commit.output)
        ));
    }

    let status = git_command::run(
        &target_dir,
        &["status", "--short", "--", ".", ":(exclude).harness"],
    );
    if status.exit_code != 0 {
        return Err(format!(
            "git status failed: {}",
            one_line(&status.error, &status.output)
        ));
    }
    if !status.output.trim().is_empty() {
        return Err(format!(
            "target directory still dirty after commit: {}",
            one_line(&status.output, "")
        ));
    }

    let hash = git_command::run(&target_dir, &["rev-parse", "--short", "HEAD"]);
    if hash.exit_code == 0 {
        Ok(one_line(&hash.output, "COMMIT_CREATED"))
    } else {
        Err(format!(
            "commit created, but the hash could not be read: {}",
            one_line(&hash.error, &hash.output)
        ))
    }
}

/// Resolves the handoff's target directory and rejects clearly dangerous or nonsensical
/// configurations before running `git add`/`git commit` in it. Full containment against a
/// signed policy root (capability broker) is future-phase work — this is just the RFC
/// §6.3 minimal rejection list: empty, filesystem root, the user's HOME, or the harness's
/// own install directory.
pub fn resolve_target_dir(target_dir: &str) -> Result<PathBuf, String> {
    let configured = target_dir.trim();
    if configured.is_empty() {
        return Err("target_dir empty: no target directory configured".to_string());
    }

    let cwd = std::env::current_dir().unwrap_or_default();
    let joined = cwd.join(configured);
    let resolved = joined.canonicalize().unwrap_or(joined);

    if resolved.parent().is_none() {
        return Err(format!(
            "target_dir resolves to the filesystem root: {}",
            resolved.display()
        ));
    }

    if let Ok(home) = std::env::var("HOME").or_else(|_| std::env::var("USERPROFILE")) {
        if !home.trim().is_empty() {
            let home_path = PathBuf::from(&home);
            let home_canonical = home_path.canonicalize().unwrap_or(home_path);
            if resolved == home_canonical {
                return Err(format!(
                    "target_dir resolves to the user's home directory: {}",
                    resolved.display()
                ));
            }
        }
    }

    if let Ok(exe) = std::env::current_exe() {
        if let Some(exe_dir) = exe.parent() {
            let exe_dir_canonical = exe_dir
                .canonicalize()
                .unwrap_or_else(|_| exe_dir.to_path_buf());
            if resolved == exe_dir_canonical {
                return Err(format!(
                    "target_dir resolves to the harness install directory: {}",
                    resolved.display()
                ));
            }
        }
    }

    Ok(resolved)
}

fn append_progress(
    target_dir: &Path,
    feature_id: i32,
    title: &str,
    verify_cmd: &str,
    verify_result: &str,
) -> std::io::Result<()> {
    let summary = one_line(&state(CURRENT_FEATURE_SUMMARY_KEY), "implementation completed");
    let verify = one_line(verify_result, "PASS");
    let command = if verify_cmd.trim().is_empty() {
        "the project's verify command".to_string()
    } else {
        verify_cmd.to_string()
    };
    let line = format!(
        "[{} UTC] Feature #{feature_id} - {}: {summary}. Verify with: {}. Result: {verify}",
        chrono::Utc::now().format("%Y-%m-%d %H:%M"),
        one_line(title, ""),
        one_line(&command, "")
    );

    if let Ok(existing) = std::fs::read_to_string(target_dir.join("progress.txt")) {
        let prefix = format!("Feature #{feature_id} - {}:", one_line(title, ""));
        if existing.lines().any(|l| l.contains(&prefix) && l.contains("Result:")) {
            return Ok(());
        }
    }

    use std::io::Write;
    let mut file = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(target_dir.join("progress.txt"))?;
    writeln!(file, "{line}")
}

fn commit_message(feature_id: i32, title: &str) -> String {
    let mut suffix = one_line(title, "");
    if suffix.len() > 72 {
        suffix = truncate_utf8_bytes(&suffix, 72).trim_end().to_string();
    }
    format!("feat(development): complete feature #{feature_id} - {suffix}")
}

/// Cuts `text` at no more than `max_bytes` UTF-8 octets, backing off to the nearest valid
/// char boundary — never splits a multi-byte character (accent, emoji) in half. Shared by
/// `commit_message` and `verify::snippet` (RFC Appendix B item 1: measure in bytes, not
/// codepoints, so the semantics match across the .NET/Python/Rust engines).
pub fn truncate_utf8_bytes(text: &str, max_bytes: usize) -> String {
    if text.len() <= max_bytes {
        return text.to_string();
    }

    let mut cut = max_bytes;
    while cut > 0 && !text.is_char_boundary(cut) {
        cut -= 1;
    }

    text[..cut].to_string()
}

pub fn one_line(value: &str, fallback: &str) -> String {
    let normalized = value
        .replace('\r', " ")
        .replace('\n', " ")
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    if normalized.trim().is_empty() {
        fallback.to_string()
    } else {
        normalized.trim().to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn truncate_utf8_bytes_does_not_split_a_multibyte_char_in_half() {
        // "café ☕" — "é" (2 bytes) and "☕" (3 bytes) are multibyte; a naive cut at any
        // position would produce an invalid UTF-8 string (panic when slicing).
        let text = "café ☕";
        for max in 0..=text.len() {
            let truncated = truncate_utf8_bytes(text, max);
            assert!(truncated.len() <= max);
        }
    }

    #[test]
    fn commit_message_truncates_a_long_title_on_utf8_byte_boundaries() {
        let title = "café ".repeat(20); // well above 72 bytes, with an accent near the cut
        let message = commit_message(1, &title);

        assert!(message.starts_with("feat(development): complete feature #1 - "));
        // If the cut had split an "é" in half, the formatting above would already have
        // panicked while building the String — reaching this point already proves the cut
        // was valid.
    }
}

File: src/go/flowsdevelopment/handoff.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"

	engine "github.com/cooperacode/IAO/src/go/harnessengine"
)

type handoffResult struct {
	Success      bool
	Confirmation string
	Failure      string
}

func okHandoff(confirmation string) handoffResult {
	return handoffResult{Success: true, Confirmation: confirmation}
}
func failedHandoff(failure string) handoffResult { return handoffResult{Failure: failure} }

func completeVerifiedFeature(verifyResult string) string {
	handoff := tryAutomatedHandoff(verifyResult)
	if !handoff.Success {
		fmt.Fprintf(os.Stderr, "[dev] automatic handoff failed: %s\n", handoff.Failure)
		return HandoffPrompt(handoff.Failure)
	}

	fmt.Fprintf(os.Stderr, "[dev] automatic handoff completed: %s\n", handoff.Confirmation)
	if id, err := strconv.Atoi(state(currentFeatureIdKey)); err == nil {
		engine.MarkFeaturePassed(id)
	}

	if engine.AllFeaturesPassing() {
		return done()
	}
	return BearingsPrompt()
}

func tryAutomatedHandoff(verifyResult string) handoffResult {
	featureId, err := strconv.Atoi(state(currentFeatureIdKey))
	if err != nil {
		return failedHandoff("current feature missing from state.json")
	}

	title := state(currentFeatureTitleKey)
	for _, f := range engine.LoadFeatures() {
		if f.Id == featureId {
			title = f.Title
			break
		}
	}
	if strings.TrimSpace(title) == "" {
		title = fmt.Sprintf("feature #%d", featureId)
	}

	config := engine.LoadRunConfig()
	targetDir, err := resolveTargetDir(config.TargetDir)
	if err != nil {
		return failedHandoff(fmt.Sprintf("invalid target directory: %s", err))
	}

	if err := os.MkdirAll(targetDir, 0o755); err != nil {
		return failedHandoff(fmt.Sprintf("failed to update progress.txt: %s", err))
	}
	if err := appendProgress(targetDir, featureId, title, config.VerifyCmd, verifyResult); err != nil {
		return failedHandoff(fmt.Sprintf("failed to update progress.txt: %s", err))
	}

	revParse := engine.RunGitCommand(targetDir, "rev-parse", "--show-toplevel")
	if revParse.ExitCode != 0 {
		return okHandoff(fmt.Sprintf("NO_GIT: %s", oneLine(revParse.Error, "target directory is outside a Git repository")))
	}

	add := engine.RunGitCommand(targetDir, "add", "-A", "--", ".", ":(exclude).harness")
	if add.ExitCode != 0 {
		return failedHandoff(fmt.Sprintf("git add failed: %s", oneLine(add.Error, add.Output)))
	}

	diff := engine.RunGitCommand(targetDir, "diff", "--cached", "--quiet", "--", ".", ":(exclude).harness")
	if diff.ExitCode == 0 {
		head := engine.RunGitCommand(targetDir, "rev-parse", "--short", "HEAD")
		if head.ExitCode == 0 {
			return okHandoff(oneLine(head.Output, "NO_CHANGES"))
		}
		return okHandoff("NO_CHANGES")
	}
	if diff.ExitCode > 1 {
		return failedHandoff(fmt.Sprintf("git diff --cached failed: %s", oneLine(diff.Error, diff.Output)))
	}

	commit := engine.RunGitCommand(targetDir, "commit", "-m", commitMessage(featureId, title), "--", ".", ":(exclude).harness")
	if commit.ExitCode != 0 {
		return failedHandoff(fmt.Sprintf("git commit failed: %s", oneLine(commit.Error, commit.Output)))
	}

	status := engine.RunGitCommand(targetDir, "status", "--short", "--", ".", ":(exclude).harness")
	if status.ExitCode != 0 {
		return failedHandoff(fmt.Sprintf("git status failed: %s", oneLine(status.Error, status.Output)))
	}
	if strings.TrimSpace(status.Output) != "" {
		return failedHandoff(fmt.Sprintf("target directory still dirty after commit: %s", oneLine(status.Output, "")))
	}

	hash := engine.RunGitCommand(targetDir, "rev-parse", "--short", "HEAD")
	if hash.ExitCode == 0 {
		return okHandoff(oneLine(hash.Output, "COMMIT_CREATED"))
	}
	return failedHandoff(fmt.Sprintf("commit created, but the hash could not be read: %s", oneLine(hash.Error, hash.Output)))
}

// resolveTargetDir applies the minimal containment (RFC §6.3): rejects targets that should
// certainly never receive automatic commits from the agent — empty, filesystem root, the
// user's HOME, or the harness's own install directory (using the running binary's
// directory as a proxy). Full containment against a signed policy root is future-phase work
// (capability broker); this is just the RFC's minimal rejection list.
func resolveTargetDir(targetDir string) (string, error) {
	if strings.TrimSpace(targetDir) == "" {
		return "", fmt.Errorf("target_dir empty/whitespace is not a valid target directory.")
	}

	resolved, err := filepath.Abs(targetDir)
	if err != nil {
		return "", fmt.Errorf("invalid target directory: %s", err)
	}

	caseInsensitive := runtime.GOOS == "windows"

	root := filepath.VolumeName(resolved) + string(filepath.Separator)
	if pathsEqual(resolved, root, caseInsensitive) {
		return "", fmt.Errorf("target_dir resolves to the filesystem root ('%s').", resolved)
	}

	if home, err := os.UserHomeDir(); err == nil {
		if normalizedHome := normalizedOrEmpty(home); normalizedHome != "" && pathsEqual(normalized(resolved), normalizedHome, caseInsensitive) {
			return "", fmt.Errorf("target_dir resolves to the user's home directory ('%s').", resolved)
		}
	}

	if harnessBase := normalizedOrEmpty(binaryDirLocal()); harnessBase != "" && pathsEqual(normalized(resolved), harnessBase, caseInsensitive) {
		return "", fmt.Errorf("target_dir resolves to the harness install directory ('%s').", resolved)
	}

	return resolved, nil
}

func binaryDirLocal() string {
	exe, err := os.Executable()
	if err != nil {
		return ""
	}
	return filepath.Dir(exe)
}

func normalized(path string) string {
	return strings.TrimRight(path, string(filepath.Separator))
}

func normalizedOrEmpty(path string) string {
	if strings.TrimSpace(path) == "" {
		return ""
	}
	abs, err := filepath.Abs(path)
	if err != nil {
		return ""
	}
	return normalized(abs)
}

func pathsEqual(a, b string, caseInsensitive bool) bool {
	if caseInsensitive {
		return strings.EqualFold(a, b)
	}
	return a == b
}

func appendProgress(targetDir string, featureId int, title, verifyCmd, verifyResult string) error {
	summary := oneLine(state(currentFeatureSummaryKey), "implementation completed")
	verify := oneLine(verifyResult, "PASS")
	command := verifyCmd
	if strings.TrimSpace(command) == "" {
		command = "the project's verify command"
	}
	line := fmt.Sprintf("[%s UTC] Feature #%d - %s: %s. Verify with: %s. Result: %s\n",
		time.Now().UTC().Format("2006-01-02 15:04"), featureId, oneLine(title, ""), summary, oneLine(command, ""), verify)
	progressPath := filepath.Join(targetDir, "progress.txt")
	if existing, err := os.ReadFile(progressPath); err == nil {
		prefix := fmt.Sprintf("Feature #%d - %s:", featureId, oneLine(title, ""))
		for _, previous := range strings.Split(string(existing), "\n") {
			if strings.Contains(previous, prefix) && strings.Contains(previous, "Result:") {
				return nil
			}
		}
	}

	f, err := os.OpenFile(progressPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = f.WriteString(line)
	return err
}

func commitMessage(featureId int, title string) string {
	suffix := oneLine(title, "")
	if len(suffix) > 72 {
		suffix = strings.TrimRight(truncateUtf8Bytes(suffix, 72), " \t")
	}
	return fmt.Sprintf("feat(development): complete feature #%d - %s", featureId, suffix)
}

// truncateUtf8Bytes cuts text at no more than maxBytes UTF-8 octets, backing off to a valid
// leading-byte boundary — never splits a multi-byte character (accent, emoji) in half.
// Duplicated from harnessengine on purpose (unexported there): this flow layer owns its own
// copy, mirroring the .NET/Rust ports where the same helper lives once per layer.
func truncateUtf8Bytes(text string, maxBytes int) string {
	if len(text) <= maxBytes {
		return text
	}

	cut := maxBytes
	for cut > 0 && (text[cut]&0xC0) == 0x80 {
		cut--
	}

	return text[:cut]
}

func oneLine(value, fallback string) string {
	normalized := strings.Join(strings.Fields(strings.ReplaceAll(strings.ReplaceAll(value, "\r", " "), "\n", " ")), " ")
	if strings.TrimSpace(normalized) == "" {
		return fallback
	}
	return strings.TrimSpace(normalized)
}
